### PR TITLE
Use the parent class loader of WebAppClassLoader for looking up Jetty…

### PR DESCRIPTION
**Pull request by lbtc-xxx**
`JettyJMXRemover` fails to find Jetty classes using `Class.forName()`. It works when it use the parent class loader of `WebAppClassLoader` instead. The Exception happend on both of `jetty-distribution-8.1.17.v20150415` and `jetty-distribution-9.3.5.v20151012`.

And, `MBeanContainer.removeBean()` is not found on latest `jetty-distribution-9.3.5.v20151012`. I don't know how it was gone but according to https://bugs.eclipse.org/bugs/show_bug.cgi?id=423255 it can be ignored because the MBean leak issue seems fixed.

Exception:

```
ClassLoaderLeakPreventor: se.jiderhamn.classloader.leak.prevention.ClassLoaderLeakPreventor shutting down context by removing known leaks (CL: 0xa51c0d5)
java.lang.ClassNotFoundException: org.eclipse.jetty.webapp.WebAppClassLoader
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at org.eclipse.jetty.webapp.WebAppClassLoader.loadClass(WebAppClassLoader.java:421)
	at org.eclipse.jetty.webapp.WebAppClassLoader.loadClass(WebAppClassLoader.java:383)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:264)
	at se.jiderhamn.classloader.leak.prevention.ClassLoaderLeakPreventor$JettyJMXRemover.<init>(ClassLoaderLeakPreventor.java:2382)
	at se.jiderhamn.classloader.leak.prevention.ClassLoaderLeakPreventor.unregisterMBeans(ClassLoaderLeakPreventor.java:931)
	at se.jiderhamn.classloader.leak.prevention.ClassLoaderLeakPreventor.contextDestroyed(ClassLoaderLeakPreventor.java:783)
	at org.eclipse.jetty.server.handler.ContextHandler.doStop(ContextHandler.java:823)
	at org.eclipse.jetty.servlet.ServletContextHandler.doStop(ServletContextHandler.java:160)
	at org.eclipse.jetty.webapp.WebAppContext.doStop(WebAppContext.java:518)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.stop(AbstractLifeCycle.java:89)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.eclipse.jetty.jmx.ObjectMBean.invoke(ObjectMBean.java:485)
	at com.sun.jmx.interceptor.DefaultMBeanServerInterceptor.invoke(DefaultMBeanServerInterceptor.java:819)
	at com.sun.jmx.mbeanserver.JmxMBeanServer.invoke(JmxMBeanServer.java:801)
	at javax.management.remote.rmi.RMIConnectionImpl.doOperation(RMIConnectionImpl.java:1471)
	at javax.management.remote.rmi.RMIConnectionImpl.access$300(RMIConnectionImpl.java:76)
	at javax.management.remote.rmi.RMIConnectionImpl$PrivilegedOperation.run(RMIConnectionImpl.java:1312)
	at javax.management.remote.rmi.RMIConnectionImpl.doPrivilegedOperation(RMIConnectionImpl.java:1404)
	at javax.management.remote.rmi.RMIConnectionImpl.invoke(RMIConnectionImpl.java:832)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at sun.rmi.server.UnicastServerRef.dispatch(UnicastServerRef.java:323)
	at sun.rmi.transport.Transport$1.run(Transport.java:200)
	at sun.rmi.transport.Transport$1.run(Transport.java:197)
	at java.security.AccessController.doPrivileged(Native Method)
	at sun.rmi.transport.Transport.serviceCall(Transport.java:196)
	at sun.rmi.transport.tcp.TCPTransport.handleMessages(TCPTransport.java:568)
	at sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run0(TCPTransport.java:826)
	at sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.lambda$run$256(TCPTransport.java:683)
	at java.security.AccessController.doPrivileged(Native Method)
	at sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run(TCPTransport.java:682)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

### CODE REVIEW

File: src/main/java/se/jiderhamn/classloader/leak/prevention/ClassLoaderLeakPreventor.java [hunk 1]
```diff
@@ -2466,5 +2471,9 @@ boolean unregisterJettyJMXBean(ObjectName objectName) {
 		    return false;
 	    }
     }
-  }  
+
+    Class getClassFromWebApplicationParentClassLoader(String className) throws ClassNotFoundException {
+      return Class.forName(className, false, getWebApplicationClassLoader().getParent());
```
Comment (mjiderhamn) [inline src/main/java/se/jiderhamn/classloader/leak/prevention/ClassLoaderLeakPreventor.java]: Will always looking at the parent be backwards compatible, or would you want to fall back to Class.forName() if not found by parent?

**Comment by lbtc-xxx** [comment]
How about this one? or would it be better if swap the order of looking up (parent first) ?

**Comment by lbtc-xxx** [comment]
> Will always looking at the parent be backwards compatible

I think it's true but I don't know why it produces ClassNotFoundException for Class.forName() but works for the parent. may further investigation is needed.

**Comment by mjiderhamn** [comment]
This seems OK, but I do think parent first may make a bit more sense.
Do you want to tweak and/or investigate before I merge and close?

**Comment by lbtc-xxx** [comment]
I have swapped the order of looking up to parent first. Unfortunately I have no time to investigate about it further ATM, so please merge it if you are OK with it. Thanks.

**Comment by lbtc-xxx** [comment]
Just for reference: I have used IntelliJ IDEA 14.1.5 to launch Jetty server that deployed the Preventor.

**Comment by mjiderhamn** [comment]
Looking great. Thanks for the pull request!

**Comment by mjiderhamn** [comment]
Released as 1.15.1

```xml
<dependency>
  <groupId>se.jiderhamn</groupId>
  <artifactId>classloader-leak-prevention</artifactId>
  <version>1.15.1</version>
</dependency>
```

**Comment by lbtc-xxx** [comment]
Thanks :)

**Pull request merged**